freud/diffraction.pyx

```diff
@@ -263,6 +265,277 @@ cdef class StaticStructureFactorDebye(_Compute):
+            del self.thisptr
+
+    def compute(self, system, query_points=None, N_total=None, reuse_box=False, reset=True):
+        r"""Computes static structure factor.
+
+        Example for a single component system::
+
+            >>> sf = freud.diffraction.StaticStructureFactorDirect(
+                    bins=100, k_max=10, k_min=0
+                )
+            >>> sf.compute((box, points))
+
+        Example for partial mixed structure factor for multiple component
+        system AB::
+
+            >>> sf = freud.diffraction.StaticStructureFactorDirect(
+                    bins=100, k_max=10, k_min=0
+                )
+            >>> sf.compute(
+                    (box, A_points),
+                    query_points=B_points, N_total=N_particles
+                )
+
+        Args:
+            system:
+                Any object that is a valid argument to
+                :class:`freud.locality.NeighborQuery.from_system`. Note that box is
+                allowed to change when calculating trajectory average static
+                structure factor. For non-orthorhombic boxes the points are wrapped
+                into a orthorhombic box.
+            query_points ((:math:`N_{query\_points}`, 3) :class:`numpy.ndarray`, optional):
+                Query points used to calculate the partial structure factor.
+                Uses the system's points if :code:`None`. See class
+                documentation for information about the normalization of partial
+                structure factors. If :code:`None`, the full scattering is
+                computed. (Default value = :code:`None`).
+            N_total (int, optional):
+                Total number of points in the system. This is required if
+                ``query_points`` are provided. See class documentation for
+                information about the normalization of partial structure
+                factors.
+            reuse_box (bool, optional):
+                wether to reuse the box from the last call. Note that box still has to be
+                provided as part of the system. If box doesn't change this saves
+                computational time. Must be used with :code:`reset = False`.
+            reset (bool, optional):
+                Whether to erase the previously computed values before adding
+                the new computation; if False, will accumulate data (Default
+                value = True).
+        """  # noqa E501
+        if (query_points is None) != (N_total is None):
+            raise ValueError(
+                "If query_points are provided, N_total must also be provided "
+                "in order to correctly compute the normalization of the "
+                "partial structure factor."
+            )
+        if reuse_box and reset:
+            raise ValueError(
+                "The reuse_box option has to be used with reset = False"
+            )
+        # Convert box to orthorhombic and wrap points inside this new box.
```

Comment:
@DomFijan This orthorhombic conversion needs to be done in C++, or else we will end up having the Python and C++ APIs returning different results.

Can you provide further explanation on why this step is needed? It doesn't seem like this would be necessary if we're correctly computing the reciprocal space vectors that are being sampled isotropically. From my brief search in the dynasor docs, it does not appear that dynasor requires orthorhombic boxes. If I missed something, let me know!

Reply:
@bdice I will attempt to move it sometime this week. 

So I tried computing S(k) using this method for a system that had a non-orthorombic box and then converted and wrapped it to an orthorombic box and got slightly different results. This is most likely (although I'm not 100% sure) due to the fact that our k-space grid is "orthorombic" (or rather rectangular?), so the corners of a tilted box probably don't get treated properly. I haven't checked this but if this is correct then the boxes with higher tilt might have a bigger error stemming from this...
In any case the system configuration and structural properties should be exactly the same no matter what box shape is chosen? By converting to an orthorombic box we sort of stick to something that we know works correctly. At least that was my thought process behind this.

Reply:
> our k-space grid is "orthorombic" (or rather rectangular?)

I don't think that's true. The k-space grid is based on multiples of the reciprocal lattice vectors (b1, b2, b3) which should correctly handle triclinic boxes. The k vectors aren't necessarily aligned with the (x, y, z) axes if there are nonzero tilt factors. See [implementation here](https://github.com/glotzerlab/freud/blob/4dfec608d641323190f7bf7f32d69c5d57cb1837/cpp/diffraction/StaticStructureFactorDirect.cc#L207-L226).

As we ran into with the diffraction pattern calculation, wrapping the points into an orthorhombic box will change the distances between some pairs of points.

As for "slightly different results" -- are you confident that the orthorhombic wrapped results were "right" and the previous implementation was "wrong"? I don't think the statement "structural properties should be exactly the same no matter what box shape is chosen" is true unless you have proof otherwise.

If we want a definitive test of this, we can re-run the dynasor sample code for liquid aluminum and use a box with tilt factors, then compare results. I think that dynasor supports arbitrary box matrices (which is even less restrictive than freud) and my previous implementation (without wrapping) should match dynasor's results.

Reply:
I totally forgot that k-space grid generator works in that way. You are right! Perhaps there shouldn't be an issue there.
Just to be sure I'll remove the conversion to orthorombic and try to test this again and come back with some concrete plots and numbers and test against dynasor as well. Perhaps I made a mistake there as well.

I cannot see though how choice of the box can influence the configuration (structure) of a snapshot of a system if one has periodic boundaries. Consider a snapshot of an infinite system of particles periodic in all directions with a given periodicity. Now attempt to tesselate the space with a shape that obeys the periodicity. Any shape that obeys the periodicity will produce the same system configuration. See the image for an example. The key here is that Lx and Ly parameters obey the periodicity in both box cases.
![boxes](https://user-images.githubusercontent.com/50439291/134225892-398914ca-da6c-41fa-91ac-de4cdb33d3ff.png)
 
When it comes to simulations, trajectories in different box types could produce different results. But this is a consequence of the fact that system might be relaxing more slowly (or might not be able to relax at all) to equilibrium due to strain as you suggested.

Reply:
@DomFijan Distances are not preserved when shearing, even if Lx, Ly are the same. In the example below, I use a system with two points. I use a square box and a sheared box. The points are the same after wrapping (they are within both boxes). However, the distances between the points under periodic boundary conventions are not the same. In the square system, the distance is `0.707...` while in the sheared case the distance is `1.118...`. I expect this difference in the minimum vectors to affect S(q) but could be convinced otherwise by a test as you proposed.

```python
import freud
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

box_square = freud.Box.square(3)
box_shear = freud.Box(3, 3, xy=0.5, is2D=True)

points = np.array([[-1.5, -1.5, 0], [-1, 1, 0]])
square_points = box_square.wrap(points)
shear_points = box_shear.wrap(points)

print(box_square)
print(box_shear)

print(points)
print(square_points)
print(shear_points)

print(box_square.compute_all_distances(points, points))
print(box_shear.compute_all_distances(points, points))

system_square = freud.AABBQuery.from_system((box_square, square_points))
system_shear = freud.AABBQuery.from_system((box_shear, shear_points))

fig, ax = plt.subplots()
system_square.plot(ax=ax)
system_shear.plot(ax=ax)
fig.savefig("result.png")
```

![result](https://user-images.githubusercontent.com/3943761/134234578-995fa9c2-c400-473c-bd6f-6be21d45b67c.png)

Reply:
You are right. Actually my argument doesn't even hold for the figure I posted. It took me good 25 mins of staring at it to realize that. Symmetry is satisfied only in x direction, while it doesn't work for y  😆 - because of the shear. 
I will be rolling back the changes sometime today or tomorrow.